docs/LemMinX-Extensions.md

```diff
@@ -101,6 +101,71 @@ The [LemMinx Extensions API](https://github.com/eclipse/lemminx/tree/master/org.
 - Formatter with [IFormatterParticipant](https://github.com/eclipse/lemminx/blob/master/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/format/IFormatterParticipant.java)
 - Symbols with [ISymbolsProviderParticipant](https://github.com/eclipse/lemminx/blob/master/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/ISymbolsProviderParticipant.java)
 
+## XML Language Server services available for extensions
+XML Language Server extension may need to use standard Language Server features such as commands, documents and ability to manipulate documents. These are available to extensions indirectly via specialized service API.
+
+#### Command Service
+The command service is a wrapper around Language Server client functionality allowing to register/unregister/execute commands via Language Server Protocol. The service can be accessed like the following:
+
+```java
+public class FooPlugin implements IXMLExtension {
+
+	@Override
+	public void start(InitializeParams params, XMLExtensionsRegistry registry) {
+		// Register here custom server command
+		IXMLCommandService commandService = registry.getCommandService();
+		commandService.registerCommand("my-cmd", (params, cancelChecker) -> "executed");		
+	}
+
+	@Override
+	public void stop(XMLExtensionsRegistry registry) {
+		// Unregister here custom server command
+		IXMLCommandService commandService = registry.getCommandService();
+		commandService.unregisterCommand("my-cmd");		
+	}
+}
+```
+
+There are two types of commands XML LS allows extender to work with via `IXMLCommandService`:
+ - Server command that can be executed from the client (via `workspace/executeCommand` request message from the LSP spec)
+ - Client command that can be executed from the server (via `xml/executeClientCommand` request message - XML extension of the LSP spec)
+ 
```

Comment:
I would add:

A server command should implement the [IDelegateCommandHandler](link to IDelegateCommandHandler):

```java
public interface IDelegateCommandHandler {

	/**
	 * Executes a command
	 * 
	 * @param params        command execution parameters
	 * @param cancelChecker check if cancel has been requested
	 * @return the result of the command
	 * @throws Exception the unhandled exception will be wrapped in
	 *                   <code>org.eclipse.lsp4j.jsonrpc.ResponseErrorException</code>
	 *                   and be wired back to the JSON-RPC protocol caller
	 */
	Object executeCommand(ExecuteCommandParams params, CancelChecker cancelChecker) throws Exception;

}

```